Run program bug fixes, vol. #2

pipping opened the pull request:
A collection of fixes that does not rely on the introduction of new condition classes.

The exit code code handling is improved but still rather messy with these changes. mkcl will still return a symbol if a process dies in response to a signal, and signal 15 may lead to exit status 143 or 15. This will be addressed in a follow-up merge request.

fare commented on uiop/image.lisp:
```diff
@@ -50,8 +50,9 @@ before the image dump hooks are called and before the image is dumped.")
-    `(and serious-condition #+ccl (not ccl:process-reset))))
+  (eval-when (:compile-toplevel :execute)
+    (deftype fatal-condition ()
+      `(and serious-condition #+ccl (not ccl:process-reset)))))
```
Can you use #+clozure instead of #+ccl ?

fare commented on uiop/image.lisp:
```diff
@@ -50,8 +50,9 @@ before the image dump hooks are called and before the image is dumped.")
 
-  (deftype fatal-condition ()
-    `(and serious-condition #+ccl (not ccl:process-reset))))
+  (eval-when (:compile-toplevel :execute)
```
1- Don't forget :load-toplevel.

2- Scratch that: we're in a with-upgradability, the eval-when is already done for us!

pipping replied:
Well. I don't claim to understand the issue. What I can tell you is that Lispworks <7 will complain that the type is redefined if I don't use this `eval-when`.

fare replied:
It looks like a bug in LispWorks 6. I can't test that. Why should anyone care about LispWorks 6?

We define plenty of other conditions without problem, what's special about fatal-condition? Does LispWorks 6 define it?

pipping replied:
The free "Personal" edition of LispWorks is still stuck at 6. I wouldn't be surprised if this was (rather than due to a lack of resources) intentional to convince cash-strapped users to go with their new affordable Hobbyist edition.

fare commented:
LGTM except for "Fix redefinition warning on LispWorks <7" which is wrong.

pipping commented:
@fare What if we wrapped the deftype in `eval-when` only on Lispworks <7? Wouldn't do any harm, would it?

fare commented:
That eval-when would PREVENT definition when loading just the FASL. Exactly the wrong thing.

fare commented:
Until you tell me why this happens with fatal-condition and not with compile-condition, I object to merging this patch. All the other ones look good (but may need be rebased). I will accept a pull request from gitlab or my github.com/fare/asdf

pipping commented:
Hehe, don't worry, I'm not trying to get you to merge that patch anymore. You've convinced me that it's wrong. I'm just providing context. A grep over asdf.lisp reveals
```
  (deftype logical-pathname () nil)
  (deftype stamp () '(or real boolean)))
  (deftype fatal-condition ()
    (deftype sb-grovel-unknown-constant-condition ()
  (deftype action () '(cons operation component)) ;; a step to be performed while building
  (deftype operation-designator ()
  (deftype user-system () '(and system (satisfies user-system-p))))
```
so that indeed there has to be something special about `fatal-condition`. Yet
```
> (error 'fatal-condition)

Error: The error 'FATAL-CONDITION is not the name of a condition class.'
```

fare commented:
I don't see any reason to use their personal edition for anything else than a trial demo that it purports to be. Not worth supporting for ASDF, not worth avoiding a warning.

pipping commented:
Fair enough. I'll just remove the patch.

pipping commented:
I've now
- Cleaned up any conflicts arising from the cmu -> cmucl switch
- Removed the lispworks <6 non-fix
- Added two more clasp-specific commits.

I've also started a wiki at https://gitlab.common-lisp.net/epipping/asdf/wikis/home.

fare commented:
OK, merged.

pipping commented:
Verified that `make test` passes on linux afterwards for the following lisps:
-    abcl-1.2.0
-    abcl-1.3.3
-    allegro cl 10.0+patches
-    ccl-1.10
-    ccl-1.11
-    clisp-2.49
-    clisp-git
-    cmucl-21a
-    ecl-16.0.0
-    ecl-16.1.2
-    ecl-git
-    mkcl-git
-    sbcl-1.1.13
-    sbcl-1.3.8

Thanks for all the fish!
